src/diffusers/models/autoencoders/dc_ae.py

```diff
@@ -0,0 +1,877 @@
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        for op in self.op_list:
+            x = op(x)
+        return x
+
+
+def build_block(
+    block_type: str, in_channels: int, out_channels: int, norm: Optional[str], act: Optional[str]
+) -> nn.Module:
+    if block_type == "ResBlock":
+        assert in_channels == out_channels
+        main_block = ResBlock(
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=3,
+            stride=1,
+            use_bias=(True, False),
+            norm=(None, norm),
+            act_func=(act, None),
+        )
+        block = ResidualBlock(main_block, nn.Identity())
+    elif block_type == "EViT_GLU":
+        assert in_channels == out_channels
+        block = EfficientViTBlock(in_channels, norm=norm, act_func=act, local_module="GLUMBConv", scales=())
+    elif block_type == "EViTS5_GLU":
+        assert in_channels == out_channels
+        block = EfficientViTBlock(in_channels, norm=norm, act_func=act, local_module="GLUMBConv", scales=(5,))
+    else:
+        raise ValueError(f"block_type {block_type} is not supported")
+    return block
+
+
+def build_stage_main(
+    width: int, depth: int, block_type: str | list[str], norm: str, act: str, input_width: int
+) -> list[nn.Module]:
+    assert isinstance(block_type, str) or (isinstance(block_type, list) and depth == len(block_type))
+    stage = []
+    for d in range(depth):
+        current_block_type = block_type[d] if isinstance(block_type, list) else block_type
+        block = build_block(
+            block_type=current_block_type,
+            in_channels=width if d > 0 else input_width,
+            out_channels=width,
+            norm=norm,
+            act=act,
+        )
+        stage.append(block)
+    return stage
+
+
+def build_downsample_block(block_type: str, in_channels: int, out_channels: int, shortcut: Optional[str]) -> nn.Module:
+    if block_type == "Conv":
+        block = nn.Conv2d(
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=3,
+            stride=2,
+            padding=1,
+        )
+    elif block_type == "ConvPixelUnshuffle":
+        block = ConvPixelUnshuffleDownsample2D(
+            in_channels=in_channels, out_channels=out_channels, kernel_size=3, factor=2
+        )
+    else:
+        raise ValueError(f"block_type {block_type} is not supported for downsampling")
+    if shortcut is None:
+        pass
+    elif shortcut == "averaging":
+        shortcut_block = PixelUnshuffleChannelAveragingDownsample2D(
+            in_channels=in_channels, out_channels=out_channels, factor=2
+        )
+        block = ResidualBlock(block, shortcut_block)
+    else:
+        raise ValueError(f"shortcut {shortcut} is not supported for downsample")
+    return block
+
+
+def build_upsample_block(block_type: str, in_channels: int, out_channels: int, shortcut: Optional[str]) -> nn.Module:
```

Comment:
let's remove this and move the logic to select upsample block type to each of the individual modules that uses a upsample block

Reply:
OK